API work package description/activity comments formatting

myabc opened the pull request:
https://www.openproject.org/work_packages/13560

myabc committed:
Use HTML description in Work Package API Resource

Alias the untextilized description as rawDescription.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
[wip] Extract text formatting helpers into Modules

Extract a number of ApplicationHelper methods (namely `textilizable`
and its dependencies) into OpenProject::TextFormatting/ObjectLinking
modules. This should enable this functionality to be more easily used
outside of the Rails controller/view context – for instance, from
Grape-based API code.

Also updates WorkPackageModel to include TextFormatting, rather than
ApplicationHelper.

Note: Only RSpecs have been updated thus far.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Prune unnecessary includes in TextFormatting spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix constant name in (empty) ObjectLinking spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/api-work-package-description-formatting

myabc committed:
Make UrlHelpers soft-dependencies of TextFormatting

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@ulferts  @hschink would you mind having a look at this? since you know the codebase better than I do, you might be aware of any potential complications.

ulferts commented on lib/open_project/text_formatting.rb:
```diff
@@ -0,0 +1,408 @@
+      project = options[:project] || @project || (obj && obj.respond_to?(:project) ? obj.project : nil)
+      only_path = options.delete(:only_path) == false ? false : true
+
+      text = Redmine::WikiFormatting.to_html(Setting.text_formatting, text, :object => obj, :attribute => attr, :edit => edit) { |macro, args| exec_macro(macro, obj, args, :view => self, :edit => edit) }
```
This one will break as `exec_macro` will call render which is only defined if the module is included into a view. 

<pre>
macro (undefined method `render' for #<API::V3::WorkPackages::WorkPackageModel:0x00000107854c30>)
</pre>

ulferts replied:
Just found out that this only applies to the `timelines` macro.

ulferts replied:
@myabc and the rest of the gang agreed to show some message along the lines of "not supported in this context" if we render outside a view context.

ulferts commented:
There is one general concern I have that should not prevent us from merging this one as I like it very much but that should make us careful when doing it. I am not sure of whether some plugins patch the rendering. Nothing comes to my mind when thinking about it but either we run it through our CI with a couple of plugins or we prepare to fix it once merged and should then not merge it too late.

ulferts commented:
Just noticed that this would also have to be applied to the comments (activity) section.

myabc committed:
Add missing I18n include to TextFormatting

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
[wip] Handle macros outside Rails' view_context

This presents a more graceful notice, rather than rendering an error.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Test macro rendering in work package resource spec

* Fix additional missing soft/implicit dependencies of WorkPackageModel

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@ulferts 

> Just noticed that this would also have to be applied to the comments (activity) section.

wasn't this done in da82c963?

ulferts commented:
@myabc: Only for the attributes that where changes. But the comment (notes in AR) where not textilized there as this would require what you did in this PR.

We will have to repeat what you did in the work_package_model/representer for the activity_model/representer.

myabc committed:
Update ng templates for HTML comments, description

n.b. this disables the expandable/collapsable description pane (inc.
truncation).

Bug # 13560

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/api-work-package-description-formatting

myabc committed:
Use HTML comments in Activity API resource

Alias the untextilized comments as rawComment.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
Note: I've made HTML the default for both description and comments (and added an equivalent raw property), This is the opposite to the approach @mtakac took in da82c96.

@ulferts @jpassing @NobodysNightmare if you have a preference, please let me know.

ulferts commented:
@myabc I prefer having an html_[attribute_name] attribute so that clients might guess that they would have to render it as html and not as plaintext. 

The non html_ prefixed attribute will on the other hand be pretty useless as there is probably no client who knows how to render our "textile". @NobodysNightmare and @jpassing take on that would interest me here. They will probably have to render the html version as well because they will otherwise have unintelligible text. 

As long as we do not stick to a standard with the raw text the HTML version is the only standard we follow. A client might be able to strip the tags and come up with plaintext. We might also be able to but that would remove the styling.

So in the end we might remove the raw version output form the API for now and stick to alway returning HTML removing the necessity of having html_\* prefixed attributes at all. But we might want to keep them if we ever plan to have something intelligible without some magic by the client.

And @myabc just pointed out correctly that we will need the client to understand the "textile" if he is to correctly update a work package. However they will do that... So my final verdict is:
- html_[attribute_name] for the html version
- [attribute_name] for textile version

jpassing commented:
I do not really care about the naming -- but it is important to offer both, html and textile.

myabc committed:
Merge branch 'dev' into feature/api-work-package-description-formatting

myabc committed:
Rename TextFormatting#textilizable to #format_text

* deprecate_alias textilizable, textilize

In the future, other markup types will be supported.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

NobodysNightmare commented:
I think we will at least need attributes as raw (i.e. Textile) values, to be able to perform a roundtrip. Everything else is additional, e.g. HTML, plain etc.

I am not sure if my proposal conforms to the standards we are using for the API (e.g. HAL), but how about:

``` JSON
myAttribute: {
  raw: "This is *my* attribute",
  html: "This is <strong>my</strong> attribute",
  plain: "This is my attribute"
}
```

I think this would be more flexible than name-prefixes. Maybe we could even rename `raw` to `textile`.

Other topic regarding different attribute formats (and tell me if it is out-of-scope):
How does the client know that it will only be able to set `myAttribute` in its raw/textile form? And what happens if the client tries to do set e.g. `html_myAttribute`?

myabc committed:
Localise macro errors visible to end-user

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
> I am not sure if my proposal conforms to the standards we are using for the API (e.g. HAL)

I don't think the standard says much about nested properties. I quite like this approach though (and it allows a degree of flexibility in the future).

myabc commented:
@NobodysNightmare unfortunately, we had to go ahead and merge _as-is_. Your points won't be disregarded though: I'll create a new PR in the specification repo there, and we can continue discussion there.
